File: Templates/Java/String/KMP.java
//
// Code Contributor(s): Rajeev Ranjan
//

import java.util.Scanner;
import java.lang.*;
import java.io.*;

class KMP
{
	public static int[] computePrefix(String pattern)
	{
		int l = pattern.length();
		int prefix[] = new int[l];
		prefix[0] = 0;
		for(int i = 1, j = 0; i < l; i++)
		{
			while(j > 0 && pattern.charAt(i) != pattern.charAt(j))
			{
				j = prefix[j-1];
			}
			if(pattern.charAt(i) == pattern.charAt(j))
			{
				prefix[i] = ++j;
			}
			else
			{
				prefix[i] = j;
			}
		}
		return prefix;
	}

	public static void KMP(String s, String pattern)
	{
		int lenOfOrg = s.length();
		int lenOfPat = pattern.length();
		int prefixArray[] = computePrefix(pattern);
		for(int i = 0, j = 0; i<lenOfOrg; i++)
		{
			while(j > 0 && s.charAt(i) != pattern.charAt(j))
			{
				j = prefixArray[j-1];
			}
			if(s.charAt(i) == pattern.charAt(j))
			{
				j++;
			}
			if(j == lenOfPat)
			{
				System.out.println("Patten starts at " + (i-lenOfPat+1) + " ends at " + i);
				j = prefixArray[j-1];
			}
		}
	}


	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int t = sc.nextInt();
		while(t-- > 0)
		{
			String s = sc.next();
			String pattern = sc.next();
			KMP(s, pattern);
		}
	}
}

File: Templates/Python/String/KMP.py
#
# Code Contributor(s): Rajeev Ranjan
#2nd contributor Sanjeev

def prefixArray(pattern):
    i = 1
    j = 0
    prefix = [0,]
    while i < len(pattern):
        while j > 0 and pattern[i] != pattern[j]:
            j = prefix[j-1]
        if pattern[i] == pattern[j]:
            j = j + 1
            prefix.insert(i, j)
        else:
            prefix.insert(i, j)
        i = i + 1
    return prefix

def KMP(string, pattern):
    m = len(string)
    n = len(pattern)
    i, j = 0, 0
    prefix = prefixArray(pattern)
    while i < m:
        while j > 0 and string[i] != pattern[j]:
            j = prefix[j-1]
        if string[i] == pattern[j]:
            j = j + 1
        if j == n:
            print("Pattern found at index : ", i-n+1)
            j = prefix[j-1]
        i = i + 1

if __name__ == '__main__':
    string = input('Enter a sentence : ')
    pattern = input('Enter the pattern to be searched : ')
    KMP(string, pattern)


'''
Test Case 0:

INPUT

Enter a sentence : My very educated mother just showed us the nine planets.
Enter the pattern to be searched : mother

OUTPUT

Pattern found at index :  17

Test Case 1:

INPUT

Enter a sentence : Repeating the same word again and again.
Enter the pattern to be searched : again

OUTPUT
Pattern found at index :  24
Pattern found at index :  34

'''
